### ISSUE TITLE
Unselecting objects with rectangle select down to a single object can crash ipe 

### ISSUE THREAD

**Issue by loizuf** [otfried/ipe#423]
If you select a set of objects with rectangle select, then hold shift to unselect all but one of these objects and finally right-click the remaining object, ipe crashes. But there are some additional things going on. First a detailed step by step recreation.

The bug first occurred in a presentation with multiple slides and objects, but I was able to recreate it in an empty ipe file.

1) Open ipe
2) Create multiple objects (It does not seem to matter if it is references, paths or text objects)
3) Select some of them (the tool you use for this selection does not seem to be important, so either click and click again with shift, or use the rectangle inclusion or intersection select)
4) Unselect all but one of these objects
5) right click -> ipe crashes (depending on which object is the left over selected one)

Actually, I tried some settings and the crash does not happen if the left-over single object was the last object created. Last here means relative to the other objects that were selected, not necessarily overall. I'll try to illustrate with the attached ipe file.

The four points A, B, C and D are added in this order.
If one selects A, B and C and then unselects A and B, a right click will not crash ipe. C is left over. C is the newest object that was selected, even though D is newer.
If one selects A, B and C and then unselects A and C, a right click will crash ipe. B is left over. A newer object (C) was selected.
If one selects A, B and C and then unselects B and C, a right click will crash ipe. A is left over. A newer object (C) was selected
If one selects B, C and D and then unselects B and D, a right click will crash ipe. C is left over. A newer object (D) was selected
If one selects B, C and D and then unselects B and C, a right click will not crash ipe. D is left over. D is the newest object that was selected, and created.

Running ipe on the command line gives you the following error
```
/usr/share/ipe/7.2.24/lua/properties.lua:68: bad argument #2 to '__index' (string expected, got nil)
stack traceback:
	[C]: in metamethod '__index'
	/usr/share/ipe/7.2.24/lua/properties.lua:68: in method 'singlePopup'
	/usr/share/ipe/7.2.24/lua/properties.lua:47: in method 'propertiesPopup'
	/usr/share/ipe/7.2.24/lua/tools.lua:1317: in local 'r'
	/usr/share/ipe/7.2.24/lua/tools.lua:1347: in function </usr/share/ipe/7.2.24/lua/tools.lua:1321>
	[C]: in function 'mainloop'
	/usr/share/ipe/7.2.24/lua/main.lua:496: in main chunk
	[C]: in function 'require'
	[string "require "main""]:1: in main chunk

```
I can only assume, something is going on, where the rectangle selection tool does not properly remove objects from some array, which holds the selected items. Or somehow the fact that there is no secondary selection anymore is causing this.

I hope I am explaining this in an understandable way.

Cheers,
Soeren

[rectselect.pdf](https://github.com/otfried/ipe-issues/files/8361493/rectselect.pdf)

**Comment by otfried** [comment]
I cannot reproduce this at all.   Do you have any non-trivial customization that could be related?

You could run Ipe after renaming your ~/.ipe/ipelets folder to something else (assuming that is where your customization is).

**Comment by loizuf** [comment]
Very curious. After changing my .ipe/ipelets folder to a different name, the crash still occurs. I have change some things in the prefs.lua file, but nothing that seems connected to this.

I also did ask two colleagues to perform the same sequence on their machine before posting (it actually happened first on the machine of a colleague) and it crashed for them too. I think they did not have any changes to their prefs files.

All three machines were running Ubuntu though (21.10 in my case and some version of 20 in their case).

**Comment by otfried** [comment]
I tried checking out out the 7.2.24 version and compiling that, but I still don't get the error.

From the error message, it seems that the page no longer has a primary selection.   When you get the error, does Ipe highlight the one remaining selected object with the primary selection color?    I was not able to trick Ipe into losing the primary selection.   You unselect the objects simply by Shift-Left Click in Select mode?

**Comment by otfried** [comment]
Hah, I got it to happen by unselecting using Shift and dragging a rectangle.   Is that what you did?

**Comment by loizuf** [comment]
Yes, I should have been more specific. Unselecting MUST happen with the rectangle (box) select and I only know one method of removing something from the current selection, while keeping the rest, which is using SHIFT.

I don't think it matters if the initial selection process was done with the rectangle select or by holding down shift and clicking. But unselecting by holding down SHIFT and clicking, instead of using the rectangle select, did NOT crash ipe.

**Comment by otfried** [comment]
Thanks, now everything is clear.  Apparently when I wrote the rectangle-selecting code, I did not think about that one can use it to only unselect objects.

Fixed in 7.2.25.
